output/sarif.go

```diff
@@ -0,0 +1,496 @@
+package output
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/open-policy-agent/opa/tester"
+)
+
+const (
+	// SARIF schema and version
+	sarifSchemaURI = "https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/schemas/sarif-schema-2.1.0.json"
+	sarifVersion   = "2.1.0"
+
+	// Tool information
+	toolName = "conftest"
+	toolURI  = "https://github.com/open-policy-agent/conftest"
+
+	// SARIF levels
+	levelError   = "error"
+	levelWarning = "warning"
+	levelNote    = "note"
+	levelNone    = "none"
+
+	// SARIF kinds
+	kindFail    = "fail"
+	kindReview  = "review"
+	kindInfo    = "informational"
+	kindPass    = "pass"
+	kindSkipped = "notApplicable"
+
+	// Rule ID prefixes
+	ruleFailureBase   = "conftest-failure"
+	ruleWarningBase   = "conftest-warning"
+	ruleExceptionBase = "conftest-exception"
+	rulePassBase      = "conftest-pass"
+	ruleSkippedBase   = "conftest-skipped"
+
+	// Result descriptions
+	successDesc   = "Policy was satisfied successfully"
+	skippedDesc   = "Policy check was skipped"
+	failureDesc   = "Policy violation"
+	warningDesc   = "Policy warning"
+	exceptionDesc = "Policy exception"
+
+	// Exit code descriptions
+	exitNoViolations = "No policy violations found"
+	exitViolations   = "Policy violations found"
+	exitWarnings     = "Policy warnings found"
+	exitExceptions   = "Policy exceptions found"
+)
+
+// SARIF represents an Outputter that outputs results in SARIF format.
+// https://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html
+type SARIF struct {
+	writer io.Writer
+}
+
+// NewSARIF creates a new SARIF with the given writer.
+func NewSARIF(w io.Writer) *SARIF {
+	return &SARIF{
+		writer: w,
+	}
+}
+
+// sarifReport represents the root object of a SARIF log file
+type sarifReport struct {
```

Comment:
Is there a need for us to maintain code for SARIF format? I am not a SME about the SARIF format, so I am hesitant to take on the maintenance of this if changes are needed in the future.

 https://github.com/owenrumney/go-sarif seems to be well maintained, would it make sense to use that instead?

Reply:
As said in another comment, appreciate the input and I've now update the PR to use go-sarif instead. I had a local schema validation kind of working, but instead of pushing it to conftest I might contribute it to go-sarif instead. Then later on also use it in conftest.